src/db/spanner/models.rs

```diff
@@ -783,6 +799,172 @@ impl SpannerDb {
+        if let Some(result) = result {
+            let total_bytes = if self.quota_enabled {
+                result[0]
+                    .get_string_value()
+                    .parse::<i64>()
+                    .map_err(|e| DbErrorKind::Integrity(e.to_string()))?
+            } else {
+                0
+            };
+            let count = result[1]
+                .get_string_value()
+                .parse::<i32>()
+                .map_err(|e| DbErrorKind::Integrity(e.to_string()))?;
+            Ok(results::GetQuotaUsage { total_bytes, count })
+        } else {
+            Ok(results::GetQuotaUsage::default())
+        }
+    }
+
+    pub async fn touch_user_collections(
+        &self,
+        user: &HawkIdentifier,
+        collection_id: i32,
+    ) -> Result<()> {
+        // This will also update the counts in user_collections, since `update_collection_sync`
+        // is called very early and only updates the timestamp.
+        // This will also write the tombstone if there are no records and we're explicitly
+        // specifying a TOMBSTONE collection_id.
+        // This function should be called after any write operation.
+        let timestamp = self.timestamp()?;
+        let mut sqlparams = params! {
+            "fxa_uid" => user.fxa_uid.clone(),
+            "fxa_kid" => user.fxa_kid.clone(),
+            "collection_id" => collection_id.to_string(),
+            "modified" => timestamp.as_rfc3339()?,
+        };
+        let mut sqltypes = param_types! {
+            "modified" => TypeCode::TIMESTAMP,
+        };
+
+        self.metrics
+            .clone()
+            .start_timer("storage.quota.update_existing_totals", None);
+        let calc_sql = if self.quota_enabled {
+            "SELECT SUM(BYTE_LENGTH(payload)), COUNT(*)
+            FROM bsos
+           WHERE fxa_uid = @fxa_uid
+             AND fxa_kid = @fxa_kid
+             AND collection_id = @collection_id
+           GROUP BY fxa_uid"
+        } else {
+            "SELECT COUNT(*)
+            FROM bsos
+           WHERE fxa_uid = @fxa_uid
+             AND fxa_kid = @fxa_kid
+             AND collection_id = @collection_id
+           GROUP BY fxa_uid"
+        };
+        let result = self
+            .sql(calc_sql)?
+            .params(params! {
+                "fxa_uid" => user.fxa_uid.clone(),
+                "fxa_kid" => user.fxa_kid.clone(),
+                "collection_id" => collection_id.to_string(),
+            })
+            .execute_async(&self.conn)?
+            .one_or_none()
+            .await?;
+        let set_sql = if let Some(result) = result {
+            // Update the user_collections table to reflect current numbers.
+            // If there are BSOs, there are user_collections (or else something
+            // really bad already happened.)
+            if self.quota_enabled {
+                let mut v = Value::new();
+                v.set_string_value(result[0].get_string_value().to_owned());
```

Comment:
as_value can help here. you should be able to take the string from the result too:

```suggestion
                let mut v = as_value(result[0].take_string_value());
```

Reply:
Ah! `take_string_value()` that's why it wasn't resolving correctly for me.